skill_predictor.py

```diff
@@ -0,0 +1,210 @@
+        """
+        avg_bag_size = 0
+        num_bags = 0
+        for user, bags in self.user_skills.items():
+            if len(bags) <= 1:
+                continue
+            for bag in bags.values():
+                if len(bag) > 2:
+                    avg_bag_size += len(bag)
+                    num_bags += 1
+        return 'Average bag size: ' + str(avg_bag_size / num_bags) + '; Number bags: ' + str(num_bags)
+
+    def _generate_skill_acquisition_times(self):
+        self.skill_acquisition_times = collections.defaultdict(list)
+        for user, time_stamps in self.user_skills.items():
+            entries = list(time_stamps.keys())
+            if len(entries) <= 1:
+                continue
+            for i in range(1, len(entries)):
+                curr_skills = time_stamps[entries[i]]
+                prev_skills = time_stamps[entries[i-1]]
+                curr_day = datetime.datetime.fromtimestamp(entries[i] / 1000)
+                prev_day = datetime.datetime.fromtimestamp(entries[i-1] / 1000)
+                day_diff = (curr_day - prev_day).total_seconds() / \
+                    (60 * 60 * 24)
+                for skill in curr_skills:
+                    if skill not in prev_skills:
+                        for prev_skill in prev_skills:
+                            self.skill_acquisition_times[(
+                                prev_skill, skill)].append(day_diff)
+
+            # for transitions with multiple records, take avg
+            for transition, time in self.skill_acquisition_times.items():
+                if len(time) > 1:
+                    self.skill_acquisition_times[transition] = [
+                        sum(time) / len(time)]
+
+    def get_skill_acquisition_times(self, skill1, skill2):
+        """
+        Finds the average acquisition time for a given skill transition
+        """
+        if self.skill_acquisition_times is None:
+            self._generate_skill_acquisition_times()
+        entry = (skill1, skill2)
+        if entry not in self.skill_acquisition_times:
+            print('No previous record of this acquisition')
+            return
+        return str(round(self.skill_acquisition_times[entry][0], 3)) + ' days'
+
+    def generate_data(self, num_permutations=1000):
+        """
+        Generates the data necessary for KenLM. Attempts to look over skill additions and deletions in a dataset and generates such changes as sentences
+        """
+        log = 0
+        # remove previous instances of files
+        try:
+            os.remove('train_data.txt')
+            os.remove('test_data.txt')
+        except OSError as e:
+            pass
+
+        for user, time_stamps in self.user_skills.items():
+            # starting data generation
+            if log % 1000 == 0:
+                print(str(log) + ' entries completed out of ' +
+                      str(len(self.user_skills)))
+            log += 1
```

Comment:
I think you want to increment log before th conditional print, or you will print a nonsense empty line the first time you enter the loop.